go/database/mpt/proof_test.go

```diff
@@ -0,0 +1,341 @@
+	if dirty {
+		t.Fatalf("expected node to be clean")
+	}
+	// remove a non-root node from the proof
+	for k := range totalProof.proofDb {
+		if k != rootHash {
+			delete(totalProof.proofDb, k)
+			break
+		}
+	}
+
+	if _, exists := totalProof.Extract(rootHash, address); exists {
+		t.Fatalf("proof should not exist")
+	}
+}
+
+func TestWitnessProof_Extract_CorruptedRlp_In_Proof(t *testing.T) {
+	ctrl := gomock.NewController(t)
+
+	address := common.Address{0xAB, 0xCD, 0xEF}
+
+	ctxt := newNodeContextWithConfig(t, ctrl, S5LiveConfig)
+
+	desc := &Extension{
+		path: AddressToNibblePath(address, ctxt)[0:30],
+		next: &Account{address: address, pathLength: 10, info: AccountInfo{Nonce: common.Nonce{0x01}, Balance: common.Balance{0x02}, CodeHash: common.Hash{0x03}}},
+	}
+
+	root, node := ctxt.Build(desc)
+	totalProof := testCreateShadowProof(t, ctxt, &root, node)
+	handle := node.GetReadHandle()
+	defer handle.Release()
+	rootHash, dirty := handle.Get().GetHash()
+	if dirty {
+		t.Fatalf("expected node to be clean")
+	}
+	// corrupt non-root node in the proof
+	for k := range totalProof.proofDb {
+		if k != rootHash {
+			totalProof.proofDb[k] = []byte{0xAA, 0xBB, 0xCC, 0xDD}
+			break
+		}
+	}
+
+	if _, exists := totalProof.Extract(rootHash, address); exists {
+		t.Fatalf("proof should not exist")
+	}
+}
+
+func TestWitnessProof_Extract_EmbeddedNode_In_Proof(t *testing.T) {
+	ctrl := gomock.NewController(t)
+
+	address := common.Address{0xAB, 0xCD, 0xEF}
+	key := common.Key{0x12, 0x34, 0x56, 0x78}
+	var value common.Value
+	value[20] = 0x02
+	value[21] = 0x04
+
+	ctxt := newNodeContextWithConfig(t, ctrl, S5LiveConfig)
+
+	desc := &Tag{label: "A", nested: &Extension{
+		path: AddressToNibblePath(address, ctxt)[0:30],
+		next: &Tag{label: "B", nested: &Account{address: address, pathLength: 34, info: AccountInfo{Nonce: common.Nonce{0x01}, Balance: common.Balance{0x02}, CodeHash: common.Hash{0x03}},
+			storage: &Tag{label: "C", nested: &Extension{
+				path:         KeyToNibblePath(key, ctxt)[0:40],
+				nextEmbedded: true,
+				next:         &Value{key: key, length: 24, value: value},
+			}}},
+		}}}
+
+	_, node := ctxt.Build(desc)
+	// proof excludes the embedded node
+	totalProof := testCreateShadowProofForLabels(t, ctxt, "A", "B", "C")
+	handle := node.GetReadHandle()
+	defer handle.Release()
+
+	rootHash, dirty := handle.Get().GetHash()
+	if dirty {
+		t.Fatalf("expected node to be clean")
+	}
+
+	proof, exists := totalProof.Extract(rootHash, address, key)
+	if !exists {
+		t.Errorf("proof should exist")
+	}
+	if got, want := proof, totalProof; !got.Equals(want) {
+		t.Errorf("unexpected proof: got %v, want %v", got, want)
+	}
+}
+
+// testCreateShadowProof creates a shadow witness proof for the given root node.
+// The shadow proof is created simply that all nodes in the MPT subtree are stored in the proof.
+// Only Empty nodes are excluded.
+func testCreateShadowProofForLabels(t *testing.T, ctxt *nodeContext, labels ...string) WitnessProof {
```

Comment:
nit: rename to `createShadowProofForLabels` or `createReferenceProofForLabels`?

The `test` prefix seems unnecessary and I initially read it as it would test the creation of that proof ..

Reply:
done